### PR TITLE
[NEAT-589] 🤓 Relationship as edge optimization

### PR DESCRIPTION
**Pull request by doctrino**
## Before
image

## After
image

Comment, the batching seems to have no effect. What helped is the caching of the entity lookup.

To get batching to work I think we need to switch to something like: https://pyoxigraph.readthedocs.io/en/stable/store.html#pyoxigraph.Store.bulk_extend

### CODE REVIEW

File: cognite/neat/_graph/extractors/_classic_cdf/_classic.py [hunk 1]
```diff
@@ -109,6 +109,18 @@ def __init__(
         self._target_external_ids_by_type: dict[InstanceIdPrefix, set[str]] = defaultdict(set)
         self._labels: set[str] = set()
         self._data_set_ids: set[int] = set()
+        self._extracted_labels = False
+        self._extracted_data_sets = False
+
+    def _get_activity_names(self) -> list[str]:
```
Comment (nikokaoja) [inline cognite/neat/_graph/extractors/_classic_cdf/_classic.py:115]: what is an activity ?
Are these all the steps/actions == activities that will be executed by this extractor?
or all the activities that took place?

File: cognite/neat/_graph/extractors/_classic_cdf/_classic.py [hunk 2]
```diff
@@ -109,6 +109,18 @@ def __init__(
...
+        activities = [core_node.extractor_cls.__name__ for core_node in self._classic_node_types] + [
```
Comment (nikokaoja) [inline cognite/neat/_graph/extractors/_classic_cdf/_classic.py:116]:
```suggestion
        activities = [dao.extractor_cls.__name__ for dao in self._classic_node_types] + [
```

Comment (nikokaoja) [reply cognite/neat/_graph/extractors/_classic_cdf/_classic.py:116]: _ClassicCoreType is misleading name for class, and I am aware that I did review 3 months ago where it slipped my attention, where in reality it is sort of a Data Access Object class, with API endpoint + config how to create triples

Comment (nikokaoja) [reply cognite/neat/_graph/extractors/_classic_cdf/_classic.py:116]: dao == Data Access Object

File: cognite/neat/_graph/extractors/_base.py [hunk 3]
```diff
@@ -10,6 +10,12 @@ class BaseExtractor:
     extractors must implement.
     """
 
+    def _get_activity_names(self) -> list[str]:
+        """Returns the name of the activities that the extractor performs."""
```
Comment (nikokaoja) [inline cognite/neat/_graph/extractors/_base.py:14]: or that performed? or will perform?

File: cognite/neat/_graph/extractors/_classic_cdf/_classic.py [hunk 4]
```diff
@@ -122,11 +134,15 @@ def extract(self) -> Iterable[Triple]:
             yield from self._extract_labels()
         except CogniteAPIError as e:
             warnings.warn(CDFAuthWarning("extract labels", str(e)), stacklevel=2)
+        else:
+            self._extracted_labels = True
```
Comment (nikokaoja) [inline cognite/neat/_graph/extractors/_classic_cdf/_classic.py:138]: I see that self._extracted_labels is set to True, meaning `_get_activity_names` should never be called before `self.extract` is called?

File: cognite/neat/_graph/transformers/_classic_cdf.py [hunk 5]
```diff
@@ -266,9 +273,9 @@ def __init__(
         {"sourceExternalId", "targetExternalId", "externalId", "sourceType", "targetType"}
     )
     _RELATIONSHIP_NODE_TYPES: tuple[str, ...] = tuple(["Asset", "Event", "File", "Sequence", "TimeSeries"])
-    description = "Replaces relationships with a schema"
+    description = "Relationships converted to edges"
```
Comment (nikokaoja) [inline cognite/neat/_graph/transformers/_classic_cdf.py:276]:
```suggestion
    description = "Converts relationships to edges"
```

File: cognite/neat/_graph/transformers/_classic_cdf.py [hunk 6]
```diff
@@ -295,7 +302,22 @@ def __init__(
     ?entity classic:externalId "{external_id}" .
 }}"""
 
+    @staticmethod
+    def create_lookup_entity(graph: Graph, namespace: Namespace) -> Callable[[str, str], URIRef]:
+        @lru_cache(maxsize=10_000)
+        def lookup_entity(entity_type: str, external_id: str) -> URIRef:
+            query = RelationshipAsEdgeTransformer._lookup_entity_query.format(
```
Comment (nikokaoja) [inline cognite/neat/_graph/transformers/_classic_cdf.py:309]: we should add classic namespace to [default_prefixes](https://github.com/cognitedata/neat/blob/5afc4a2444b6a5106abd63f91aad8b6c372e9bb5/cognite/neat/_constants.py#L74) and make sure they are all registered when [upserting prefixes in store](https://github.com/cognitedata/neat/blob/5afc4a2444b6a5106abd63f91aad8b6c372e9bb5/cognite/neat/_store/_base.py#L99)

File: cognite/neat/_graph/transformers/_classic_cdf.py [hunk 7]
```diff
@@ -295,7 +302,22 @@ def __init__(
     ?entity classic:externalId "{external_id}" .
 }}"""
 
+    @staticmethod
+    def create_lookup_entity(graph: Graph, namespace: Namespace) -> Callable[[str, str], URIRef]:
+        @lru_cache(maxsize=10_000)
```
Comment (nikokaoja) [inline cognite/neat/_graph/transformers/_classic_cdf.py:307]: I am not sure why would this help as the sparql query is executed anyhow yielding you all results back

Comment (nikokaoja) [reply cognite/neat/_graph/transformers/_classic_cdf.py:307]: what do you mean when you say `lookup_entity` ? you are getting rdf:type Relationship but only those that have `classix:externalId` as property?

File: cognite/neat/_graph/transformers/_classic_cdf.py [hunk 8]
```diff
@@ -305,45 +327,70 @@ def transform(self, graph: Graph) -> None:
...
+    def _edge_triples(
+        self,
+        graph: Graph,
+        source_type: str,
+        target_type: str,
+        instance_count: int,
+        entity_lookup: Callable[[str, str], URIRef],
+    ) -> Iterable[Triple]:
+        query = self._instances.format(namespace=self._namespace, source_type=source_type, target_type=target_type)
+        total = instance_count if self._limit_per_type is None else self._limit_per_type
```
Comment (nikokaoja) [inline cognite/neat/_graph/transformers/_classic_cdf.py:342]: naming is confusing here, what total means?

Comment (nikokaoja) [reply cognite/neat/_graph/transformers/_classic_cdf.py:342]: and why is it falling back to self._limit_per_type ?
Is self._limit_per_type set somewhere?

File: cognite/neat/_graph/transformers/_classic_cdf.py [hunk 9]
```diff
@@ -305,45 +327,70 @@ def transform(self, graph: Graph) -> None:
...
+
+        converted_relationships: list[URIRef] = []
+        for no, result in enumerate(
```
Comment (nikokaoja) [inline cognite/neat/_graph/transformers/_classic_cdf.py:345]: so here you are stepping through various edge types Asset2Asset, Asset2Timeseries, etc.

File: cognite/neat/_graph/transformers/_classic_cdf.py [hunk 10]
```diff
@@ -305,45 +327,70 @@ def transform(self, graph: Graph) -> None:
...
+            iterate_progress_bar(graph.query(query), total=total, description="Relationships to edges")
+        ):
+            if self._limit_per_type is not None and no >= self._limit_per_type:
```
Comment (nikokaoja) [inline cognite/neat/_graph/transformers/_classic_cdf.py:348]: you break if there is more than _limit_per_type ? is _limit_per_type upper or lower bound?

File: cognite/neat/_graph/transformers/_classic_cdf.py (hunk 7 above)
Comment (nikokaoja) [reply cognite/neat/_graph/transformers/_classic_cdf.py:307]: Name `create_lookup_entity` is not concrete, what you are actually doing here is to get ID of instance by supplying value for its externalid property. Again my mistake not throughly reviewing your previous PR.

**Review by nikokaoja — APPROVED** [review]
I will approve it but def will take a stab next week to optimize it and make a bit more readable as this is a bit convoluted.

File: cognite/neat/_graph/extractors/_classic_cdf/_classic.py (hunk 4 above)
Comment (doctrino) [reply cognite/neat/_graph/extractors/_classic_cdf/_classic.py:138]: That is correct.

File: cognite/neat/_graph/transformers/_classic_cdf.py (hunk 7 above)
Comment (doctrino) [reply cognite/neat/_graph/transformers/_classic_cdf.py:307]: This is new, so you have not seen it before. Renamed it now.

File: cognite/neat/_graph/transformers/_classic_cdf.py (hunk 10 above)
Comment (doctrino) [reply cognite/neat/_graph/transformers/_classic_cdf.py:348]: Upper bound. It is to now convert absolutely every relationship in the graph.

**Pull request merged**